xds/src/main/java/io/grpc/xds/AbstractXdsClient.java

```diff
@@ -0,0 +1,760 @@
+      // used for management server to identify which response the client is ACKing/NACking.
+      // To avoid confusion, client-initiated requests will always use the nonce in
+      // most recently received responses of each resource type.
+      switch (type) {
+        case LDS:
+          ldsRespNonce = nonce;
+          handleLdsResponse(versionInfo, resources, nonce);
+          break;
+        case RDS:
+          rdsRespNonce = nonce;
+          handleRdsResponse(versionInfo, resources, nonce);
+          break;
+        case CDS:
+          cdsRespNonce = nonce;
+          handleCdsResponse(versionInfo, resources, nonce);
+          break;
+        case EDS:
+          edsRespNonce = nonce;
+          handleEdsResponse(versionInfo, resources, nonce);
+          break;
+        case UNKNOWN:
+        default:
+          logger.log(XdsLogLevel.WARNING, "Ignore an unknown type of DiscoveryResponse");
+      }
+    }
+
+    final void handleRpcError(Throwable t) {
+      handleRpcStreamClosed(Status.fromThrowable(t));
+    }
+
+    final void handleRpcCompleted() {
+      handleRpcStreamClosed(Status.UNAVAILABLE.withDescription("Closed by server"));
+    }
+
+    private void handleRpcStreamClosed(Status error) {
+      checkArgument(!error.isOk(), "unexpected OK status");
+      if (closed) {
+        return;
+      }
+      logger.log(
+          XdsLogLevel.ERROR,
+          "ADS stream closed with status {0}: {1}. Cause: {2}",
+          error.getCode(), error.getDescription(), error.getCause());
+      closed = true;
+      handleStreamClosed(error);
+      cleanUp();
+      if (responseReceived || retryBackoffPolicy == null) {
+        // Reset the backoff sequence if had received a response, or backoff sequence
+        // has never been initialized.
+        retryBackoffPolicy = backoffPolicyProvider.get();
+      }
+      long delayNanos = 0;
+      if (!responseReceived) {
+        delayNanos =
+            Math.max(
+                0,
+                retryBackoffPolicy.nextBackoffNanos()
+                    - stopwatch.elapsed(TimeUnit.NANOSECONDS));
+      }
+      logger.log(XdsLogLevel.INFO, "Retry ADS stream in {0} ns", delayNanos);
+      rpcRetryTimer = timeService.schedule(new RpcRetryTask(), delayNanos, TimeUnit.NANOSECONDS);
+    }
+
+    private void close(Exception error) {
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      sendError(error);
+    }
+
+    private void cleanUp() {
+      if (adsStream == this) {
+        adsStream = null;
+      }
+    }
+  }
+
+  private final class AdsStreamV2 extends AbstractAdsStream {
+    private final io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc
+        .AggregatedDiscoveryServiceStub stubV2;
+    private StreamObserver<io.envoyproxy.envoy.api.v2.DiscoveryRequest> requestWriterV2;
+
+    AdsStreamV2() {
+      stubV2 = io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc.newStub(
```

Comment:
`stubV2` can be moved into `start()` as a local variable.

Reply:
Sure. Cleaned up the `AdsStreamV2` and `AdsStreamV3` naming as well, made the class itself more descriptive.